plugins/outputs/zabbix/utils.go

```diff
@@ -0,0 +1,126 @@
+package zabbix
+
+import (
+	"fmt"
+	"os"
+	"sort"
+	"strings"
+
+	"github.com/datadope-io/go-zabbix/v2"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/internal"
+)
+
+var (
+	errMetricNoMeasurementName = fmt.Errorf("empty measurement name")
+	errMetricNoTags            = fmt.Errorf("metric without tags")
+	errMetricJustHostTag       = fmt.Errorf("metric with only host tag")
+)
+
+// getHostname returns the hostname from the tags, or the system hostname if not found.
+func getHostname(tags []*telegraf.Tag) (string, error) {
+	for _, tag := range tags {
+		if tag.Key == hostTag {
+			return tag.Value, nil
+		}
+	}
+
+	return os.Hostname()
+}
+
+// processMetric converts a Telegraf metric to a list of Zabbix metrics.
+// Ignore metrics with no hostname.
+func (z Zabbix) processMetric(metric telegraf.Metric) []*zabbix.Metric {
+	zbxMetrics := make([]*zabbix.Metric, 0, len(metric.FieldList()))
+
+	for _, field := range metric.FieldList() {
+		zbxMetric, err := z.buildZabbixMetric(metric, field.Key, field.Value)
+		if err != nil {
+			z.Log.Errorf("Error converting telegraf metric to Zabbix format: %v", err)
+			continue
+		}
+
+		zbxMetrics = append(zbxMetrics, zbxMetric)
+	}
+
+	return zbxMetrics
+}
+
+// buildZabbixMetric builds a Zabbix metric from a Telegraf metric, for one particular value.
+func (z Zabbix) buildZabbixMetric(metric telegraf.Metric, fieldName string, value interface{}) (*zabbix.Metric, error) {
+	hostname, err := getHostname(metric.TagList())
+	if err != nil {
+		return nil, fmt.Errorf("error getting hostname: %w", err)
+	}
+
+	metricValue, err := internal.ToString(value)
+	if err != nil {
+		return nil, fmt.Errorf("error converting value: %w", err)
+	}
+
+	key := z.Prefix + metric.Name() + "." + fieldName
+	if z.SkipMeasurementPrefix {
+		key = z.Prefix + fieldName
+	}
+
+	// Ignore host tag.
+	// We want to add tags to the key in alphabetical order. Eg.:
+	// lld.dns_query.query_time_ms[DOMAIN,RECORD_TYPE,SERVER]
+	tags := make([]string, 0, len(metric.TagList()))
+
+	for _, tag := range metric.TagList() {
+		if tag.Key == hostTag {
+			continue
+		}
+
+		tags = append(tags, tag.Key)
+	}
+
+	sort.Strings(tags)
+
+	if len(tags) != 0 {
+		tagValues := make([]string, 0, len(tags))
+
+		// Get tag values in the same order as the tag keys in the tags slice.
+		for _, tag := range tags {
+			for _, metricTag := range metric.TagList() {
+				if metricTag.Key == tag {
+					tagValues = append(tagValues, metricTag.Value)
+				}
+			}
```

Comment:
```suggestion
			tagValues = append(tagValues, metric.Tag(tag))
```

Reply:
Fixed the suggested change in 8a2248bd8

Reply:
Why used `GetTag` instead of simply `Tag`? As indeed, the tag will always exist..

Reply:
I thought it was an error, as `Tag(string)` does not exists.
But I guess you mean `Tag()[string]`.

Fixed in 3086adabf

Reply:
No, I really mean `Tag(string)` as it really does exist: https://github.com/influxdata/telegraf/blob/4b447d04830e4ab908a348f8446a1c125ecd106f/metric/metric.go#L178-L181

Reply:
mmm, but that is for `telegraf.metric`, not `telegraf.Metric`, right?
I'm missing something :/

Reply:
Yeah, it's part of the `telegraf.TemplateMetric` interface, but have a look at https://github.com/influxdata/telegraf/pull/13739#discussion_r1328486890 for another way into this..